src/query/api/v1/handler/prometheus/remote/write.go

```diff
@@ -155,23 +211,55 @@ func newPromWriteMetrics(scope tally.Scope) (promWriteMetrics, error) {
 	ingestLatencyBuckets = append(ingestLatencyBuckets, upTo60mBuckets...)
 	ingestLatencyBuckets = append(ingestLatencyBuckets, upTo6hBuckets...)
 	ingestLatencyBuckets = append(ingestLatencyBuckets, upTo24hBuckets...)
+
+	var forwardLatencyBuckets tally.DurationBuckets
+	forwardLatencyBuckets = append(forwardLatencyBuckets, upTo1sBuckets...)
+	forwardLatencyBuckets = append(forwardLatencyBuckets, upTo10sBuckets...)
+	forwardLatencyBuckets = append(forwardLatencyBuckets, upTo60sBuckets...)
+	forwardLatencyBuckets = append(forwardLatencyBuckets, upTo60mBuckets...)
 	return promWriteMetrics{
 		writeSuccess:         scope.SubScope("write").Counter("success"),
 		writeErrorsServer:    scope.SubScope("write").Tagged(map[string]string{"code": "5XX"}).Counter("errors"),
 		writeErrorsClient:    scope.SubScope("write").Tagged(map[string]string{"code": "4XX"}).Counter("errors"),
 		ingestLatency:        scope.SubScope("ingest").Histogram("latency", ingestLatencyBuckets),
 		ingestLatencyBuckets: ingestLatencyBuckets,
+		forwardSuccess:       scope.SubScope("forward").Counter("success"),
+		forwardErrors:        scope.SubScope("forward").Counter("errors"),
+		forwardDropped:       scope.SubScope("forward").Counter("dropped"),
+		forwardLatency:       scope.SubScope("forward").Histogram("latency", forwardLatencyBuckets),
 	}, nil
 }
 
 func (h *PromWriteHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	req, opts, rErr := h.parseRequest(r)
+	req, opts, result, rErr := h.parseRequest(r)
 	if rErr != nil {
 		h.metrics.writeErrorsClient.Inc(1)
 		xhttp.Error(w, rErr.Inner(), rErr.Code())
 		return
 	}
 
+	// Kick of forwarding.
+	// NB(r): Be careful about not returning buffers to pool
+	// if the request bodies ever get pooled until after
+	// forwarding completes.
+	if targets := h.forwarding.Targets; len(targets) > 0 {
+		for _, target := range targets {
+			target := target
+			spawned := h.forwardingWorkerPool.GoIfAvailable(func() {
+				if err := h.forward(r.Context(), result, target); err != nil {
```

Comment:
Is this safe? Not sure what happens if the top level request completes and closes the context before this goroutine has a chance to execute.

Reply:
There's no pooling, so the context will already be cancelled.

Reply:
I'll explicitly take assignment of the ctx before this call rather than calling `r.Context()` to make it more clear.